### PR TITLE
jonmv/document and test dynamic throttling policy

### PR DESCRIPTION
**Pull request by jonmv**
@baldersheim and @vekterli see if this makes sense to you :) The first commit is actually a bugfix from the previous PR 😬

### CODE REVIEW

File: messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java [hunk 1]
```diff
@@ -8,11 +8,43 @@
...
 /**
  * This is an implementation of the {@link ThrottlePolicy} that offers dynamic limits to the number of pending messages a
- * {@link SourceSession} is allowed to have.
- *
- * <b>NOTE:</b> By context, "pending" is referring to the number of sent messages that have not been replied to yet.
+ * {@link SourceSession} is allowed to have. Pending means the number of sent messages that have not been replied to yet.
+ * <p>
+ * The algorithm works by increasing the number of messages allowed to be pending, the <em>winidow size</em>, until
+ * this no longer increases throughput. At this point, the algorithm is driven by synthetic attraction towards latencies
+ * which satisfy <code>log10(1 / latency) % 1 = e</code>, for some constant <code>0 < e < 1</code>. Weird? Most certainly!
```
Comment (jonmv) [inline messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java:15]:
```suggestion
 * which satisfy <code>log10(1 / latency) % 1 = e</code>, for some constant <code>0 &lt; e &lt; 1</code>. Weird? Most certainly!
```

**Comment by jonmv** [comment]
The unit tests are random. Could set seeds to ensure consistent results, but I think limits generous enough that shouldn't be necessary. With fixed seeds, we could just be very lucky, and the test has less value imo.

**Review by vekterli — APPROVED** [review]
> Any sufficiently advanced technology is indistinguishable from magic.

🦄

I'm not smart enough to analyze this for correctness, but infinitely more tests than before is always good! (But see rogue `println`)

File: messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java [hunk 2]
```diff
@@ -8,11 +8,43 @@
 
 /**
  * This is an implementation of the {@link ThrottlePolicy} that offers dynamic limits to the number of pending messages a
- * {@link SourceSession} is allowed to have.
- *
- * <b>NOTE:</b> By context, "pending" is referring to the number of sent messages that have not been replied to yet.
+ * {@link SourceSession} is allowed to have. Pending means the number of sent messages that have not been replied to yet.
+ * <p>
+ * The algorithm works by increasing the number of messages allowed to be pending, the <em>winidow size</em>, until
```
[the remark this answers is not in the document]
Comment (jonmv) [reply messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java:13]: Thanks!

File: messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java [hunk 3]
```diff
@@ -96,30 +129,31 @@ public void processMessage(Message message) {
...
+            while(throughput * period / windowSize < 2) {
                 period *= 10;
             }
-            while(throughput * period/windowSize > 2) {
+            while(throughput * period / windowSize > 2) {
                 period *= 0.1;
             }
-            double efficiency = throughput*period/windowSize;
+            double efficiency = throughput * period / windowSize; // "efficiency" is a strange name. This is where on the level it is.
+            if (Math.random() < 1e-2) System.err.println(efficiency);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java:151]: Absolutely not.

File: messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jonmv) [reply messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java:13]:
```suggestion
 * The algorithm works by increasing the number of messages allowed to be pending, the <em>window size</em>, until
```

File: messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonmv) [reply messagebus/src/main/java/com/yahoo/messagebus/DynamicThrottlePolicy.java:151]:
```suggestion
```

**Comment by jonmv** [comment]
Can't say my analysis was documented very well either 😬 But spending more time on this now feels like waste. It's a combination of expecting things from code and verifying this is so with tests. I believe the algorithm may have worked differently at some point, as well, based on its javadoc, but I'm certain that is now incorrect.

**Pull request merged**